soco/discovery.py

```diff
@@ -77,14 +92,34 @@ def discover(timeout=1, include_invisible=False, interface_addr=None):
         except socket.error:
             raise ValueError("{0} is not a valid IP address string".format(
                 interface_addr))
-        _sock.setsockopt(
-            socket.IPPROTO_IP, socket.IP_MULTICAST_IF, address)
-
-    # Send a few times. UDP is unreliable
-    _LOG.info("Sending discovery packets")
-    _sock.sendto(really_utf8(PLAYER_SEARCH), (MCAST_GRP, MCAST_PORT))
-    _sock.sendto(really_utf8(PLAYER_SEARCH), (MCAST_GRP, MCAST_PORT))
-    _sock.sendto(really_utf8(PLAYER_SEARCH), (MCAST_GRP, MCAST_PORT))
+        _sockets[0] = create_socket(interface_addr)
+    else:
+        # Find the local network address using a couple of different methods.
+        # Create a socket for each unique address found, and one for the
+        # default multicast address
+        addresses = set()
+        try:
+            addresses.add(socket.gethostbyname(socket.gethostname()))
+        except socket.error:
+            pass
+        try:
+            addresses.add(socket.gethostbyname(socket.getfqdn()))
+        except socket.error:
+            pass
+        for address in addresses:
+            _sockets.append(create_socket(address))
+        # Add a socket using the system default address
+        _sockets.append(create_socket(None))
```

Comment:
> With the way that this is appended - could we end up with one address being duplicated, as it is no longer adding it to a set, but an array

You are right. Packets may end up being sent out on the same interface twice, since we do not know (and cannot tell) what the system default address is, so it may be the same as one already in the list. There is no obvious (to me) way round this. But I don't see that it matters at all. It just means that one interface gets more packets than another.